### PR TITLE
bump esp-hal to v0.21.0

### PR DESCRIPTION
**Pull request by kaspar030**
# Description

Bumps esp-hal to `0.21.0` + our patches.

## Issues/PRs references

Depends on #399.

## Change checklist

- [x] I have cleaned up my commit history and squashed fixup commits.
- [x] I have followed the [Coding Conventions](../book/src/coding-conventions.md).
- [x] I have performed a self-review of my own code.
- [ ] I have made corresponding changes to the documentation.

### CODE REVIEW

**Comment by kaspar030** [comment]
With this, WiFi now works on the `esp32-wroom-32`!

**Comment by kaspar030** [comment]
This cargo-deny error seems like a bug:

```
INFO resolve: Resolving package: 'cargo-deny'
 WARN The package cargo-deny v0.16.1 (x86_64-unknown-linux-musl) has been downloaded from github.com
 INFO This will install the following binaries:
 INFO   - cargo-deny (cargo-deny -> /home/runner/.cargo/bin/cargo-deny)
 INFO Installing binaries...
 INFO Done in 1.838378756s
thread '<unnamed>' panicked at /home/runner/.cargo/registry/src/index.crates.io-6f17d22bba15001f/krates-0.17.1/src/builder.rs:718:25:
internal error: entered unreachable code: unable to locate defmt-03 for crate registry+https://github.com/rust-lang/crates.io-index#embedded-hal@0.2.7 features(["unproven"])
note: run with `RUST_BACKTRACE=1` environment variable to display a backtrace
Error: Process completed with exit code 101.
```

**Comment by kaspar030** [comment]
I've tested `examples/embassy-http-server` with the c6, plain esp32 and the s3. s3 goes into a boot loop, but also does so on main.

File: src/riot-rs-esp/src/i2c/controller/mod.rs [hunk 1]
```diff
@@ -109,12 +108,11 @@ macro_rules! define_i2c_drivers {
                     // does not seem possible to disable the timeout feature on ESP; so we keep the
                     // default timeout instead (encoded as `None`).
                     let timeout = None;
-                    let twim = I2C::new_with_timeout_async(
+                    let twim = EspI2c::new_with_timeout_async(
                         i2c_peripheral,
                         sda_pin,
                         scl_pin,
                         frequency,
-                        &clocks,
```
[the remark this answers is not in the document]
Comment (kaspar030) [reply src/riot-rs-esp/src/i2c/controller/mod.rs:117]: yeah! the esp-hal team does a great job refining things IMO.

**Comment by elenaf9** [comment]
Unfortunately this PR broke threading for `xtensa`, not sure why. Both threading examples don't work anymore on main, but still worked on the commit before this PR.

**Pull request merged**